File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def solution(s):
    answer = 0
    s = list(s)
    for _ in range(len(s)):
        stack = []
        for i in s:
            if not stack:
                stack.append(i)
            else:
                top = stack[-1]
                if top == '(' and i == ')':
                    stack.pop()
                elif top == '[' and i == ']':
                    stack.pop()
                elif top == '{' and i == '}':
                    stack.pop()
                else:
                    stack.append(i)
        if not stack:
            answer += 1
        s.append(s[0])
        s = s[1:]
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        Deque<Character> deq = new ArrayDeque<>();
        for (char c : s.toCharArray()) {
            deq.offerLast(c);
        }
        if(check(deq)) answer++;
        for (int i = 1; i < s.length(); i++) {
            deq.offerLast(deq.pollFirst());
            if(check(deq)) answer++;
        }
        return answer;
    }
    
    public boolean check(Deque<Character> dq) {
        Deque<Character> stk = new ArrayDeque<>();
        for (char c : dq) {
            if (c == '(' || c == '[' || c == '{') {
                stk.push(c);
            } else {
                if (stk.isEmpty()) return false;

                char top = stk.pop();
                if ((c == ')' && top != '(') ||
                    (c == ']' && top != '[') ||
                    (c == '}' && top != '{')) {
                    return false;
                }
            }
        }

        return stk.isEmpty();
    }
}